apps/api/src/app/workflows-v2/util/template-parser/new-liquid-parser.ts

```diff
@@ -0,0 +1,947 @@
+      if (token) {
+        addVariable(
+          {
+            name: extractVariableFromOutput(token.input) || 'unknown',
+            message: error.message,
+            output: token.input,
+            outputStart: token.begin,
+            outputEnd: token.end,
+          },
+          false
+        );
+      }
+    }
+  }
+
+  return { validVariables, invalidVariables };
+}
+
+function parseByLiquid({
+  template,
+  variableSchema,
+}: {
+  template: string;
+  variableSchema?: JSONSchemaDto;
+}): TemplateVariables {
+  const validVariables: Array<Variable> = [];
+  const invalidVariables: Array<Variable> = [];
+  const parsed = parserEngine.parse(template);
+
+  processTemplates({
+    templates: parsed,
+    validVariables,
+    invalidVariables,
+    variableSchema,
+    localVariables: new Set(),
+  });
+
+  return { validVariables, invalidVariables };
+}
+
+function processTemplates(context: ProcessContext) {
+  const { templates, validVariables, invalidVariables, variableSchema, localVariables = new Set() } = context;
+
+  templates.forEach((template: Template) => {
+    if (isOutputToken(template)) {
+      processOutputToken({
+        template,
+        validVariables,
+        invalidVariables,
+        variableSchema,
+        localVariables,
+      });
+    } else if (isTagToken(template)) {
+      processTagToken({
+        template,
+        validVariables,
+        invalidVariables,
+        variableSchema,
+        localVariables,
+      });
+    }
+  });
+}
+
+function isPropertyAllowed(schema: JSONSchemaDto | undefined, propertyPath: string) {
+  if (!schema) {
+    return true;
+  }
+
+  let currentSchema = { ...schema };
+  if (typeof currentSchema !== 'object') {
+    return false;
+  }
+
+  const pathParts = propertyPath
+    .split('.')
+    .map((part) => {
+      // Split array notation into [propName, index]
+      const arrayMatch = part.match(/^(.+?)\[(\d+)\]$/);
+
+      return arrayMatch ? [arrayMatch[1], arrayMatch[2]] : [part];
+    })
+    .flat();
+
+  for (const part of pathParts) {
+    const { properties, additionalProperties, type } = currentSchema;
+
+    // Handle direct property access
+    if (properties?.[part]) {
+      currentSchema = properties[part] as JSONSchemaDto;
+      continue;
+    }
+
+    // Handle array paths - valid if schema is array type
+    if (type === 'array') {
+      // Valid array index or property access
+      const isArrayIndex = !Number.isNaN(Number(part)) && Number(part) >= 0;
+      const arrayItemSchema = currentSchema.items as Record<string, unknown>;
+
+      if (isArrayIndex) {
+        currentSchema = arrayItemSchema;
+        continue;
+      }
+
+      if (arrayItemSchema?.properties?.[part]) {
+        currentSchema = arrayItemSchema.properties[part];
+        continue;
+      }
+    }
+
+    if (additionalProperties === true) {
+      return true;
+    }
+
+    return false;
+  }
+
+  return true;
+}
+
+function validateVariable({
```

Comment:
validate variable against the local variables (in the context of loop) or schema